R/driver_rds_remote.R

```diff
@@ -0,0 +1,199 @@
+        write_if_missing("TRUE", path =  storr:::driver_rds_config_file(path, "compress"), file_ops = file_ops)
+        write_if_missing("md5", path =  storr:::driver_rds_config_file(path, "hash_algorithm"), file_ops = file_ops)
+      }
+      ## Then write out the version number:
+      write_if_missing(value = as.character(packageVersion("storr")),
+                       path = storr:::driver_rds_config_file(path, "version"),
+                       file_ops = file_ops)
+
+      if (!is.null(mangle_key)) {
+        storr:::assert_scalar_logical(mangle_key)
+      }
+      self$mangle_key <- driver_rds_config(file_ops, path, "mangle_key", mangle_key,
+                                           FALSE, TRUE)
+
+      if (!is.null(mangle_key_pad)) {
+        storr:::assert_scalar_logical(mangle_key_pad)
+      }
+      self$mangle_key_pad <-driver_rds_config(file_ops, path, "mangle_key_pad",
+                                              mangle_key_pad, FALSE, TRUE)
+
+      if (!is.null(compress)) {
+        storr:::assert_scalar_logical(compress)
+      }
+      self$compress <- driver_rds_config(file_ops, path, "compress", compress,
+                                         TRUE, FALSE)
+
+      if (!is.null(hash_algorithm)) {
+        storr:::assert_scalar_character(hash_algorithm)
+      }
+      self$hash_algorithm <- driver_rds_config(file_ops, path, "hash_algorithm",
+                                               hash_algorithm, "md5", TRUE)
+    },
+
+    type = function() {
+      "s3"
+    },
+
+    destroy = function() {
+      self$file_ops$delete_recursive(path = self$path)
+    },
+
+    get_hash = function(key, namespace) {
+      self$file_ops$readLines(path = self$name_key(key, namespace))
+    },
+
+    set_hash = function(key, namespace, hash) {
+      self$file_ops$create_dir(path = self$name_key("", namespace))
+      self$file_ops$writeLines(text = hash, path = self$name_key(key, namespace))
+      #*** should be making use of (or making an equivalent version of) the
+      #write_lines function within the storr package here (I think it deletes
+      #file if the write fails)
+    },
+
+    get_object = function(hash) {
+      self$file_ops$readObject(path = self$name_hash(hash))
+    },
+
+    set_object = function(hash, value) {
+      ## NOTE: this takes advantage of having the serialized value
+      ## already and avoids seralising twice.
+      storr:::assert_raw(value)
+
+      self$file_ops$write_serialized_rds(value = value,
+                                         filename = self$name_hash(hash),
+                                         compress = self$compress)
+    },
+
+    exists_hash = function(key, namespace) {
+      self$file_ops$object_exists(self$name_key(key, namespace))
+    },
+
+    exists_object = function(hash) {
+      self$file_ops$object_exists(self$name_hash(hash))
+    },
+
+    del_hash = function(key, namespace) {
```

Comment:
There's no need to use an _actual_ directory.  In the redis driver there is no directory either - it's just a schema over the keys.  So there I use (I think) something like `data:<hash>` and `keys:<namespace>:<key>` - you could do the same thing here storing things as `data_<hash>.rds` and `keys_namespace_value` - though you have to work out what you'll do with names that include an underscore!  Alternatively, you could use `keys/namespace/value` but then base64 encode that and store that as a file on S3

Reply:
the only place where this gets really nasty is deletion because you lose the ability to remove a directory recursively and you can get race conditions with a list/del pair

Reply:
I did think about not using an actual directory, but (I think) that would mean an entire S3 bucket would need to be dedicated to each store, and so i leaned in favor of trying to mimic a directory structure. This does mean that multiple storrs can be created in a bucket, and you can browse the file structure using an S3 client. I tried to create delete functions that work, but I'm not sure a good strategy for testing them out.

Reply:
It'd be great if this could be documented somewhere (the layout on S3)